File: sidecar/protocol.go
// Package sidecar defines the process boundary between the engine and a model
// that is not written in Go.
//
// Many audio-model inference stacks live in Python. Letting that Python into
// the build, the test path, or the analysis path would make every one of them
// slower and more fragile, so it stays on the other side of a process
// boundary with a documented, versioned protocol between. The conformance
// suite is the contract: a sidecar that passes it works, whatever it is
// written in, and a sidecar that does not is broken before anyone runs a
// model.
//
// # Framing
//
// Each message is one JSON header line terminated by a newline, optionally
// followed by exactly PayloadBytes of binary payload:
//
//	{"type":"audio","payload_bytes":960}\n<960 bytes of PCM16>
//
// Audio is the only thing large enough to matter and it is raw rather than
// base64, because a third more bytes and an encode/decode pass on every frame
// is a real cost on the hot path. Everything else is ordinary JSON, so a
// sidecar can be debugged by reading the stream.
package sidecar

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/bojieli/OpenRealtime/element"
)

// Version is the frozen, legacy protocol version and remains the default.
// VersionInteraction adds typed interaction-act handoff. VersionMultimodal
// adds direct image frames and live tool-catalog updates. VersionElementGraph
// carries arbitrary descriptor-checked element ports rather than assuming an
// audio binding. Selecting a later version is explicit in Config so existing
// sidecars are never reinterpreted merely because the engine learned a new
// frame.
const Version = 1

const (
	VersionInteraction  = 2
	VersionMultimodal   = 3
	VersionElementGraph = 4
	LatestVersion       = VersionElementGraph
)

// MessageType names one frame.
type MessageType string

// Engine to sidecar.
const (
	// TypeHello opens a session and carries its configuration. It is always
	// the first message, and a sidecar must answer with Ready or Error.
	TypeHello MessageType = "hello"
	// TypeAudio carries input PCM16 at the negotiated rate.
	TypeAudio MessageType = "audio"
	// TypeImage carries one encoded JPEG or PNG frame. It is protocol v3: the
	// model sees pixels directly rather than a narration standing in for them.
	TypeImage MessageType = "image"
	// TypeToolsUpdate replaces the live function catalog after session.update.
	// A Realtime binding starts before the client's tool declaration arrives,
	// so a hello-only catalog would always be stale in production.
	TypeToolsUpdate MessageType = "tools_update"
	// TypeText injects text into the model's context without asking it to
	// respond. It is how a background reasoner's answer reaches a model that
	// owns its own voice.
	TypeText MessageType = "text"
	// TypeCommit closes the input audio buffer: the client has declared its
	// turn over rather than waiting for silence. A sidecar whose model owns
	// its own floor decides for itself what to do about that; one that does
	// not can ignore it, because the engine has already ended the turn.
	TypeCommit MessageType = "commit"
	// TypeRespond asks the model to produce a turn now.
	TypeRespond MessageType = "respond"
	// TypeInterrupt stops generation at the model's next safe point.
	TypeInterrupt MessageType = "interrupt"
	// TypeInteractionAct carries a policy decision rather than collapsing it
	// into respond or interrupt. It is available in protocol v2 when the
	// sidecar declares CapabilityInteractionActs.
	TypeInteractionAct MessageType = "interaction_act"
	// TypeToolResult returns the outcome of a call the model requested.
	TypeToolResult MessageType = "tool_result"
	// TypeBye ends the session cleanly.
	TypeBye MessageType = "bye"
	// TypeElementFrame carries one envelope on one descriptor-declared port.
	// It is symmetric: the engine sends frames only to input ports and the
	// sidecar sends frames only from output ports. Trigger, interrupt, timing,
	// terminal, and state semantics remain in the port's element.Type rather
	// than being collapsed into transport-specific message kinds.
	TypeElementFrame MessageType = "element_frame"
)

// Sidecar to engine.
const (
	// TypeReady answers Hello and declares what this sidecar can do.
	TypeReady MessageType = "ready"
	// TypeSpeechStarted and TypeSpeechStopped report model-native voice
	// activity. A sidecar that has no native detector never sends them, and
	// the engine keeps the floor.
	TypeSpeechStarted MessageType = "speech_started"
	TypeSpeechStopped MessageType = "speech_stopped"
	// TypeTranscript reports what the model heard.
	TypeTranscript MessageType = "transcript"
	// TypeTextDelta and TypeTextDone report what the model is saying, as text.
	TypeTextDelta MessageType = "text_delta"
	TypeTextDone  MessageType = "text_done"
	// TypeOutputAudio carries output PCM16 at the declared rate.
	TypeOutputAudio MessageType = "output_audio"
	// TypeTurnDone marks the end of one model turn.
	TypeTurnDone MessageType = "turn_done"
	// TypeToolCall reports a call the model wants made. A sidecar that cannot
	// call tools never sends it.
	TypeToolCall MessageType = "tool_call"
	// TypeError reports a failure. It does not end the session unless Fatal.
	TypeError MessageType = "error"
	// TypeLog carries a diagnostic line, so a sidecar's own logging does not
	// have to fight for stderr with the process that started it.
	TypeLog MessageType = "log"
)

// Message is one frame. Payload is carried out of band and is not part of the
// JSON; it is populated by the reader and consumed by the writer.
type Message struct {
	Type         MessageType `json:"type"`
	PayloadBytes int         `json:"payload_bytes,omitempty"`

	// Hello and Ready.
	Version      int      `json:"version,omitempty"`
	Model        string   `json:"model,omitempty"`
	SampleRate   int      `json:"sample_rate,omitempty"`
	OutputRate   int      `json:"output_rate,omitempty"`
	Instructions string   `json:"instructions,omitempty"`
	Voice        string   `json:"voice,omitempty"`
	Capabilities []string `json:"capabilities,omitempty"`
	Tools        []Tool   `json:"tools,omitempty"`

	// Generic element handshake and frames (protocol v4). The full descriptor
	// is confirmed at readiness so a package declaration cannot silently drift
	// from the process that actually started. SelectedPorts is the graph's
	// active observation/action surface, not everything the implementation is
	// theoretically capable of supporting.
	ElementDescriptor    *element.Descriptor     `json:"element_descriptor,omitempty"`
	ElementConfig        json.RawMessage         `json:"element_config,omitempty"`
	SelectedPorts        []PortSelection         `json:"selected_ports,omitempty"`
	RequiredCapabilities []CapabilityRequirement `json:"required_capabilities,omitempty"`
	RuntimeArtifact      ArtifactIdentity        `json:"runtime_artifact,omitempty"`
	ResolvedCapabilities []CapabilityIdentity    `json:"resolved_capabilities,omitempty"`
	Port                 string                  `json:"port,omitempty"`
	Envelope             *WireEnvelope           `json:"envelope,omitempty"`
	// Selected ownership (protocol v2). Capabilities say what the stack can
	// do; these fields tell a sidecar which available provider is active.
	InteractionOwner string `json:"interaction_owner,omitempty"`
	FloorOwner       string `json:"floor_owner,omitempty"`

	// Transcript, text, and log.
	Text  string `json:"text,omitempty"`
	Final bool   `json:"final,omitempty"`
	Level string `json:"level,omitempty"`

	// Text injection.
	Role string `json:"role,omitempty"`

	// Tool call and result.
	CallID    string          `json:"call_id,omitempty"`
	Name      string          `json:"name,omitempty"`
	Arguments json.RawMessage `json:"arguments,omitempty"`
	Output    json.RawMessage `json:"output,omitempty"`
	Error     string          `json:"error,omitempty"`

	// Errors.
	Code  string `json:"code,omitempty"`
	Fatal bool   `json:"fatal,omitempty"`

	// Timing.
	TimestampMS int64  `json:"timestamp_ms,omitempty"`
	Source      string `json:"source,omitempty"`
	MIMEType    string `json:"mime_type,omitempty"`
	Width       int    `json:"width,omitempty"`
	Height      int    `json:"height,omitempty"`

	// Typed interaction plan (protocol v2).
	Act         string  `json:"act,omitempty"`
	Policy      string  `json:"policy,omitempty"`
	EvidenceRef string  `json:"evidence_ref,omitempty"`
	Floor       string  `json:"floor,omitempty"`
	DeadlineMS  int64   `json:"deadline_ms,omitempty"`
	Confidence  float64 `json:"confidence,omitempty"`
	Abstained   bool    `json:"abstained,omitempty"`

	// Payload is the binary body. It is never marshalled into the header.
	Payload []byte `json:"-"`
}

// Tool is a function definition handed to a sidecar.
type Tool struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Parameters  json.RawMessage `json:"parameters"`
}

// Capability names something a sidecar can do. The engine reads these to
// decide what it must supply itself, which is what makes a partial
// implementation useful rather than broken.
type Capability string

const (
	// CapabilityNativeVAD means the model detects voice activity itself. A
	// sidecar without it gets an engine-owned floor.
	CapabilityNativeVAD Capability = "native_vad"
	// CapabilityTranscript means the model reports what it heard.
	CapabilityTranscript Capability = "transcript"
	// CapabilityTextInjection means text can be added to the model's context
	// between turns. Without it, a background reasoner cannot reach the model
	// at all, and the binding must fall back to an explicit hand-off.
	CapabilityTextInjection Capability = "text_injection"
	// CapabilityVisualInput means protocol-v3 image frames reach the model as
	// pixels. A narrated screen does not satisfy this capability.
	CapabilityVisualInput Capability = "visual_input"
	// CapabilityTools means the model can request function calls.
	CapabilityTools Capability = "tools"
	// CapabilityBargeIn means the model handles overlap itself.
	CapabilityBargeIn Capability = "barge_in"
	// CapabilityFullDuplex means the model can listen and speak at once. Floor
	// and interaction ownership are selected independently by the binding.
	CapabilityFullDuplex Capability = "full_duplex"
	// CapabilityNativeInteraction means the model chooses conversational acts
	// from its own multimodal state. It is independent of full duplex and of
	// native VAD: those are capabilities a model often has alongside it, not
	// synonyms for it.
	CapabilityNativeInteraction Capability = "native_interaction"
	// CapabilityInteractionActs means the sidecar accepts protocol-v2 typed
	// plans from an external interaction controller.
	CapabilityInteractionActs Capability = "interaction_acts"
)

// Validate rejects a malformed frame before it reaches either side.
//
// The checks are deliberately about structure rather than about semantics: a
// sidecar author debugging their implementation should get "audio needs a
// payload" rather than a silent hang three seconds later.
func (message Message) Validate() error {
	if strings.TrimSpace(string(message.Type)) == "" {
		return errors.New("a sidecar message requires a type")
	}
	if message.PayloadBytes < 0 {
		return errors.New("payload length cannot be negative")
	}
	if message.PayloadBytes != len(message.Payload) && len(message.Payload) > 0 {
		return fmt.Errorf("payload declares %d bytes and carries %d", message.PayloadBytes, len(message.Payload))
	}
	switch message.Type {
	case TypeHello:
		if message.Version == VersionElementGraph {
			return validateElementHello(message)
		}
		if message.Version <= 0 || message.SampleRate <= 0 {
			return errors.New("hello requires a version and an input sample rate")
		}
		if message.Version >= VersionInteraction {
			for field, owner := range map[string]string{
				"interaction_owner": message.InteractionOwner, "floor_owner": message.FloorOwner,
			} {
				if owner != "engine" && owner != "model" && owner != "remote" {
					return fmt.Errorf("v2 hello %s must be engine, model, or remote", field)
				}
			}
		}
	case TypeReady:
		if message.Version == VersionElementGraph {
			return validateElementReady(message)
		}
		if message.Version <= 0 || message.OutputRate <= 0 {
			return errors.New("ready requires a version and an output sample rate")
		}
		if strings.TrimSpace(message.Model) == "" {
			return errors.New("ready requires the model identity, so evidence can name what produced it")
		}
	case TypeAudio, TypeOutputAudio:
		if len(message.Payload) == 0 {
			return fmt.Errorf("%s requires a PCM16 payload", message.Type)
		}
		if len(message.Payload)%2 != 0 {
			return fmt.Errorf("%s payload must contain whole PCM16 samples", message.Type)
		}
	case TypeImage:
		if len(message.Payload) == 0 || strings.TrimSpace(message.Source) == "" ||
			message.Width <= 0 || message.Height <= 0 {
			return errors.New("image requires encoded bytes, a source, and positive geometry")
		}
		if message.MIMEType != "image/jpeg" && message.MIMEType != "image/png" {
			return fmt.Errorf("image MIME type must be image/jpeg or image/png, got %q", message.MIMEType)
		}
	case TypeText:
		if strings.TrimSpace(message.Text) == "" {
			return errors.New("text injection requires text")
		}
	case TypeTranscript, TypeTextDelta:
		if strings.TrimSpace(message.Text) == "" {
			return fmt.Errorf("%s requires text", message.Type)
		}
	case TypeToolCall:
		if strings.TrimSpace(message.CallID) == "" || strings.TrimSpace(message.Name) == "" {
			return errors.New("a tool call requires a call ID and a name")
		}
		if len(message.Arguments) == 0 || !json.Valid(message.Arguments) {
			return errors.New("tool call arguments must be valid JSON")
		}
	case TypeToolResult:
		if strings.TrimSpace(message.CallID) == "" {
			return errors.New("a tool result requires a call ID")
		}
		hasOutput := len(message.Output) > 0
		hasError := strings.TrimSpace(message.Error) != ""
		if hasOutput == hasError {
			return errors.New("a tool result requires exactly one of output or error")
		}
	case TypeInteractionAct:
		validActs := map[string]string{
			"listen": "unchanged", "speak-through": "preserve", "answer": "take",
			"interrupt": "take", "act-silently": "unchanged",
			"keep-speaking": "unchanged", "stop-speaking": "yield",
		}
		floor, valid := validActs[strings.TrimSpace(message.Act)]
		if !valid {
			return fmt.Errorf("interaction act %q is not in the protocol vocabulary", message.Act)
		}
		if message.Floor != floor {
			return fmt.Errorf("interaction act %q requires floor %q, got %q", message.Act, floor, message.Floor)
		}
		if strings.TrimSpace(message.Policy) == "" || strings.TrimSpace(message.EvidenceRef) == "" {
			return errors.New("an interaction act requires policy and evidence identities")
		}
		if message.DeadlineMS <= 0 {
			return errors.New("an interaction act requires a positive deadline")
		}
		if message.Confidence < 0 || message.Confidence > 1 {
			return errors.New("interaction confidence must be between zero and one")
		}
	case TypeError:
		if strings.TrimSpace(message.Message()) == "" {
			return errors.New("an error requires a message")
		}
	case TypeElementFrame:
		return validateElementFrame(message)
	case TypeCommit, TypeRespond, TypeInterrupt, TypeBye, TypeSpeechStarted, TypeSpeechStopped,
		TypeTextDone, TypeTurnDone, TypeLog, TypeToolsUpdate:
	default:
		return fmt.Errorf("unknown sidecar message type %q", message.Type)
	}
	return nil
}

// Message returns an error frame's human-readable text.
func (message Message) Message() string {
	if message.Text != "" {
		return message.Text
	}
	return message.Error
}

// Has reports whether a ready frame declared a capability.
func (message Message) Has(capability Capability) bool {
	for _, declared := range message.Capabilities {
		if declared == string(capability) {
			return true
		}
	}
	return false
}

File: sidecars/openrealtime_sidecar/protocol.py
"""Framing for the OpenRealtime sidecar protocol.

Each message is one JSON header line terminated by a newline, optionally
followed by exactly ``payload_bytes`` of binary payload::

    {"type":"audio","payload_bytes":960}\\n<960 bytes of PCM16>

Legacy media frames and protocol-v4 typed element frames may use the binary
lane.  In v4 the selected port's explicit wire format, not an audio-specific
message name, owns the JSON/binary lanes and their bounds.  Headers remain
ordinary JSON so the boundary is inspectable without decoding provider data.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field
from typing import Any, BinaryIO

# Version 1 remains accepted and is what an engine requests by default.
# Version 2 adds typed interaction-act handoff. Version 3 adds direct encoded
# images and tool-catalog updates. Version 4 carries descriptor-checked element
# ports and is deliberately selected by the engine rather than inferred.
VERSION = 4
SUPPORTED_VERSIONS = (1, 2, 3, 4)

MAX_HEADER_BYTES = 1 << 20
MAX_PAYLOAD_BYTES = 16 << 20


class MessageType:
    """Every frame name in the protocol."""

    # Engine to sidecar.
    HELLO = "hello"
    AUDIO = "audio"
    IMAGE = "image"
    TOOLS_UPDATE = "tools_update"
    TEXT = "text"
    COMMIT = "commit"
    RESPOND = "respond"
    INTERRUPT = "interrupt"
    INTERACTION_ACT = "interaction_act"
    TOOL_RESULT = "tool_result"
    BYE = "bye"
    ELEMENT_FRAME = "element_frame"

    # Sidecar to engine.
    READY = "ready"
    SPEECH_STARTED = "speech_started"
    SPEECH_STOPPED = "speech_stopped"
    TRANSCRIPT = "transcript"
    TEXT_DELTA = "text_delta"
    TEXT_DONE = "text_done"
    OUTPUT_AUDIO = "output_audio"
    TURN_DONE = "turn_done"
    TOOL_CALL = "tool_call"
    ERROR = "error"
    LOG = "log"


class Capability:
    """What a sidecar declares it can do.

    The engine reads these to decide what it must supply itself, which is what
    makes a partial implementation useful rather than broken.
    """

    NATIVE_VAD = "native_vad"
    TRANSCRIPT = "transcript"
    TEXT_INJECTION = "text_injection"
    TOOLS = "tools"
    BARGE_IN = "barge_in"
    FULL_DUPLEX = "full_duplex"
    NATIVE_INTERACTION = "native_interaction"
    INTERACTION_ACTS = "interaction_acts"
    VISUAL_INPUT = "visual_input"


@dataclass
class Message:
    """One protocol frame."""

    type: str
    header: dict[str, Any] = field(default_factory=dict)
    payload: bytes = b""
    # Exact header bytes are retained for protocol-v4 attestations and byte
    # limits. Parsing JSON loses harmless-but-byte-significant number spelling
    # and whitespace, so reconstructing a digest from Python values would not
    # prove what actually crossed the process boundary.
    raw_header: bytes = b""

    def get(self, name: str, default: Any = None) -> Any:
        return self.header.get(name, default)

    @property
    def text(self) -> str:
        return str(self.header.get("text", ""))

    @property
    def final(self) -> bool:
        return bool(self.header.get("final", False))


def _reject_duplicate_fields(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for name, value in pairs:
        if name in result:
            raise ValueError(f"sidecar header repeats field {name!r}")
        result[name] = value
    return result


def _reject_non_finite_number(value: str) -> Any:
    raise ValueError(f"sidecar header contains non-finite number {value}")


def _read_exact(stream: BinaryIO, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining:
        chunk = stream.read(remaining)
        if not chunk:
            carried = length - remaining
            raise ValueError(f"payload declared {length} bytes and carried {carried}")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _write_all(stream: BinaryIO, source: bytes) -> None:
    remaining = memoryview(source)
    while remaining:
        written = stream.write(remaining)
        if written is None or written <= 0:
            raise OSError("short write while emitting a sidecar frame")
        remaining = remaining[written:]


def read_message(stream: BinaryIO) -> Message | None:
    """Read one frame, or None at the end of the stream."""
    # Bound the read itself. Checking only after an unbounded readline lets a
    # malicious peer allocate arbitrary memory before the limit is applied.
    line = stream.readline(MAX_HEADER_BYTES + 1)
    if not line:
        return None
    if len(line) > MAX_HEADER_BYTES:
        raise ValueError("sidecar header exceeds the frame limit")
    if not line.endswith(b"\n"):
        raise ValueError("sidecar header is not newline-terminated")
    try:
        header = json.loads(
            line,
            object_pairs_hook=_reject_duplicate_fields,
            parse_constant=_reject_non_finite_number,
        )
    except (TypeError, json.JSONDecodeError) as failure:
        raise ValueError(f"invalid sidecar header: {failure}") from failure
    if not isinstance(header, dict):
        raise ValueError("sidecar header must be one JSON object")
    message_type = header.get("type", "")
    if not isinstance(message_type, str) or not message_type.strip() or message_type != message_type.strip():
        raise ValueError("a sidecar message requires a canonical type")
    declared = header.get("payload_bytes", 0)
    if isinstance(declared, bool) or not isinstance(declared, int):
        raise ValueError("payload length must be an integer")
    payload_bytes = declared
    if payload_bytes < 0 or payload_bytes > MAX_PAYLOAD_BYTES:
        raise ValueError(f"payload length {payload_bytes} is out of range")
    payload = b""
    if payload_bytes:
        # A payload that does not follow its declared length would
        # desynchronise the stream permanently, so it fails here rather than
        # three frames later. BinaryIO.read is allowed to return short reads.
        payload = _read_exact(stream, payload_bytes)
    return Message(
        type=message_type, header=header, payload=payload, raw_header=line[:-1],
    )


def write_message(
    stream: BinaryIO, message_type: str, payload: bytes = b"", **header: Any
) -> None:
    """Write one frame and flush it.

    Flushing every frame is deliberate. A sidecar that buffers is a sidecar
    that appears to hang, and the whole point of streaming audio is that it
    arrives while it is still useful.
    """
    if not isinstance(message_type, str) or not message_type.strip() or message_type != message_type.strip():
        raise ValueError("a sidecar message requires a canonical type")
    if not isinstance(payload, bytes):
        raise TypeError("sidecar payload must be bytes")
    if len(payload) > MAX_PAYLOAD_BYTES:
        raise ValueError(f"sidecar payload exceeds the {MAX_PAYLOAD_BYTES}-byte frame limit")
    header = {name: value for name, value in header.items() if value is not None}
    header["type"] = message_type
    header.pop("payload_bytes", None)
    if payload:
        header["payload_bytes"] = len(payload)
    encoded = json.dumps(
        header, separators=(",", ":"), ensure_ascii=False, allow_nan=False,
    ).encode("utf-8")
    if len(encoded) + 1 > MAX_HEADER_BYTES:
        raise ValueError("sidecar header exceeds the frame limit")
    _write_all(stream, encoded + b"\n")
    if payload:
        _write_all(stream, payload)
    stream.flush()


def log(message: str) -> None:
    """Write a diagnostic line to stderr.

    The engine forwards it. A traceback that reaches the operator is the
    difference between "the model failed" and knowing why.
    """
    print(message, file=sys.stderr, flush=True)
